core/common/lib/policy-engine-lib/src/main/java/org/eclipse/edc/policy/engine/plan/PolicyEvaluationPlanner.java

```diff
@@ -0,0 +1,324 @@
+            planner = new PolicyEvaluationPlanner(scope);
+        }
+
+        public static PolicyEvaluationPlanner.Builder newInstance(String scope) {
+            return new PolicyEvaluationPlanner.Builder(scope);
+        }
+
+        public Builder ruleValidator(RuleValidator ruleValidator) {
+            planner.ruleValidator = ruleValidator;
+            return this;
+        }
+
+        public Builder preValidator(String scope, BiFunction<Policy, PolicyContext, Boolean> validator) {
+
+            if (scopeFilter(scope, planner.delimitedScope)) {
+                planner.preValidators.add(validator);
+            }
+
+            return this;
+        }
+
+        public Builder preValidators(String scope, List<BiFunction<Policy, PolicyContext, Boolean>> validators) {
+            validators.forEach(validator -> preValidator(scope, validator));
+            return this;
+        }
+
+        public Builder postValidator(String scope, BiFunction<Policy, PolicyContext, Boolean> validator) {
+            if (scopeFilter(scope, planner.delimitedScope)) {
+                planner.postValidators.add(validator);
+            }
+            return this;
+        }
+
+        public Builder postValidators(String scope, List<BiFunction<Policy, PolicyContext, Boolean>> validators) {
+            validators.forEach(validator -> postValidator(scope, validator));
+            return this;
+        }
+
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        public <R extends Rule> Builder evaluationFunction(String scope, String key, Class<R> ruleKind, AtomicConstraintFunction<R> function) {
+
+            if (scopeFilter(scope, planner.delimitedScope)) {
+                planner.constraintFunctions.computeIfAbsent(key, k -> new ArrayList<>())
+                        .add(new ConstraintFunctionEntry(ruleKind, function));
+            }
+            return this;
+        }
+
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        public <R extends Rule> Builder evaluationFunction(String scope, Class<R> ruleKind, DynamicAtomicConstraintFunction<R> function) {
+            if (scopeFilter(scope, planner.delimitedScope)) {
+                planner.dynamicConstraintFunctions.add(new DynamicAtomicConstraintFunctionEntry(ruleKind, function));
+            }
+            return this;
+        }
+
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        public <R extends Rule> Builder evaluationFunction(String scope, Class<R> ruleKind, RuleFunction<R> function) {
+            if (scopeFilter(scope, planner.delimitedScope)) {
+                planner.ruleFunctions.add(new RuleFunctionFunctionEntry(ruleKind, function));
+            }
+            return this;
+        }
+
+        public PolicyEvaluationPlanner build() {
+            Objects.requireNonNull(planner.ruleValidator, "Rule validator should not be null");
+            return planner;
+        }
+
+        private boolean scopeFilter(String entry, String scope) {
```

Comment:
this function is the same as the one in `PolicyEngineImpl`, it could be extracted into a static method/function